sys/include/auto_init_utils.h

```diff
@@ -0,0 +1,125 @@
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_auto_init_utils Utilities to influence the order of auto-initialized modules
+ * @ingroup     sys
+ *
+ * @{
+ * @file
+ * @brief       Auto-initialization utilities
+ *
+ * A module priority value in RIOTs auto-initialization sequence is, as of now,
+ * a hard coded 4-digit number, but is likely to be changed and automatically
+ * generated in the future. Modules are initialized in ascending
+ * (alpha)numerical order of their priority, which means that modules with
+ * lower priorities are initialized first, as long as their priorities comply
+ * with the current rules. The left most digit is reserved for the top level
+ * group the modules is situated in. The next two digits are used to enumerate
+ * the modules within their group. Hence, the rightmost digit is reserved to
+ * insert up to 9 user defined modules in between two RIOT internal modules.
+ * To test the experimental auto-init sequence, enable
+ * CONFIG_AUTO_INIT_PRIO_EXPERIMENTAL in Kconfig.
+ *
+ * @experimental
+ * @author      Fabian Hüßler <fabian.huessler@ovgu.de>
+ */
+#ifndef AUTO_INIT_UTILS_H
+#define AUTO_INIT_UTILS_H
+
+#include <stdio.h>
+#include "xfa.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef CONFIG_AUTO_INIT_ENABLE_DEBUG
+/**
+ * @brief   Print a debug message before a module is initialized
+ */
+#define CONFIG_AUTO_INIT_ENABLE_DEBUG 0
+#endif
+
+#if !defined(CONCAT) || defined(DOXYGEN)
+/**
+ * @brief   @ref CONCAT helper macro
+ * @internal
+ */
+#define CONCAT_HELP(a, b) a ## b
+/**
+ * @brief   Classic concat macro
+ * @internal
+ */
+#define CONCAT(a, b) CONCAT_HELP(a, b)
+#endif
+#if !defined(EVAL) || defined(DOXYGEN)
+/**
+ * @brief   @ref EVAL helper macro
+ * @internal
+ */
+#define EVAL_HELP(a) a
+/**
+ * @brief   Classic eval macro
+ * @internal
+ */
+#define EVAL(a) EVAL_HELP(a)
+#endif
+#if !defined(STRINGIFY) || defined(DOXYGEN)
+/**
+ * @brief   @ref STRINGIFY helper macro
+ * @internal
+ */
+#define STRINGIFY_HELP(a) #a
+/**
+ * @brief   Classic stringify macro
+ * @internal
+ */
+#define STRINGIFY(a) STRINGIFY_HELP(a)
+#endif
+
+/**
+ * @brief   Auto-init function type
+ */
+typedef void (*auto_init_fn_t)(void);
+
+/**
+ * @brief   Type to represent a module to be auto-initialized
+ */
+typedef struct {
+    auto_init_fn_t init;    /**< function to initialize the module */
+} auto_init_module_t;
+
+/**
+ * @brief   Add a module to the auto-initialization array
+ *
+ * @param   module      Module name
+ * @param   prio        Priority level
+ * @param   function    Function to be called on initialization
+ * @param   ...         String format and arguments to be printed with @ref DEBUG()
+ */
+#define AUTO_INIT_MODULE(module, prio, function, ...)                                   \
+    static inline void CONCAT(CONCAT(_debug_, prio), function)(void) {                  \
```

Comment:
To my taste, there's too much macro magic. A macro-generated function calling out to printf? Not a fan.

How about we add fields to auto_init_module_t, like, module_name, prio, extra. "extra" could be a user supplied `const char *` (string literal), so multiple instances of the same module can be differentiated.
The printing can then be done in plain C in `auto_init()`.

That way we could drop *all* macros from this file but `AUTO_INIT_MODULE` (assuming XTSTR is re-used from macros/xtstr.h, for setting the module name field from "module" arg).

Reply:
Ok, but do we always want to pay for these struct members. I guess only if debug is enabled. Or should each member get its own Kconfig flag? Like `CONFIG_AUTO_INIT_MODULE_HAS_NAME`, `CONFIG_AUTO_INIT_MODULE_HAS_PRIO`, `CONFIG_AUTO_INIT_MODULE_HAS_EXTRA`.

Reply:
yeah, `#ifdef AUTOINIT_DEBUG` (or whatever) only